cmd/clusterctl/client/cluster/upgrader.go

```diff
@@ -174,6 +187,37 @@ func (u *providerUpgrader) ApplyCustomPlan(coreProvider clusterctlv1.Provider, u
 	return u.doUpgrade(upgradePlan)
 }
 
+// ensureSingleProviderInstance ensures only one instance of a provider is running, returns error otherwise.
+func (u *providerUpgrader) ensureSingleProviderInstance() error {
+	providers, err := u.providerInventory.List()
+	if err != nil {
+		return err
+	}
+
+	providerTypeGrouping := make(map[string][]string)
+	for _, p := range providers.Items {
+		namespacedName := types.NamespacedName{Namespace: p.Namespace, Name: p.Name}.String()
+		if providers, ok := providerTypeGrouping[p.Type]; ok {
+			providerTypeGrouping[p.Type] = append(providers, namespacedName)
+		} else {
+			providerTypeGrouping[p.Type] = []string{namespacedName}
+		}
+	}
+
+	var errs []error
+	for pType, pNames := range providerTypeGrouping {
+		if len(pNames) > 1 {
+			errs = append(errs, errors.Errorf("multiple instance of provider type %q found: %v", pType, pNames))
```

Comment:
Given that this error is going to surface up to the user I think we should give a little bit more context and possibly link to the page in the book where we are explaining multi tenancy